waiter/src/waiter/scheduler/shell.clj

```diff
@@ -694,12 +806,20 @@
         get-service->instances-fn #(get-service->instances id->service-agent)
         {:keys [retrieve-syncer-state-fn]}
         (start-scheduler-syncer-fn scheduler-name get-service->instances-fn scheduler-state-chan scheduler-syncer-interval-secs)
-        {:keys [id->service-agent port->reservation-atom] :as scheduler}
+        {:keys [id->service-agent port->reservation-atom work-directory] :as scheduler}
         (create-shell-scheduler (assoc config
                                   :id->service-agent id->service-agent
                                   :retrieve-syncer-state-fn retrieve-syncer-state-fn))
         http-client (http/client {:connect-timeout health-check-timeout-ms
                                   :idle-timeout health-check-timeout-ms})]
+    (when backup-file-name
+      (let [backup-file-path (str work-directory (File/separator) backup-file-name)]
+        ;; restore the state of the shell scheduler
+        (restore-state scheduler backup-file-path)
+        ;; persist scheduler state any time the id->service-agent state changes
+        (add-watch id->service-agent ::shell-backup
```

Comment:
What is `::shell-backup`?

Reply:
It is a label associated with the watch function, it can be used to remove the watch.